pkg/healthchecker/health_checker_windows.go

```diff
@@ -0,0 +1,138 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package healthchecker
+
+import (
+	"fmt"
+	"net/http"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/node-problem-detector/cmd/healthchecker/options"
+	"k8s.io/node-problem-detector/pkg/healthchecker/types"
+	"k8s.io/node-problem-detector/pkg/util"
+)
+
+// getUptimeFunc returns the time for which the given service has been running.
+func getUptimeFunc(service string) func() (time.Duration, error) {
+	return func() (time.Duration, error) {
+		// Using the WinEvent Log Objects to find the Service logs' time when the Service was last stopped/terminated.
+		// The powershell command formats the TimeCreated of the event log in RFC1123Pattern.
+		// However, because the time library parser does not recognize the ',' in this RFC1123Pattern format,
+		// it is manually removed before parsing it using the UptimeTimeLayout.
+		getTimeCreatedCmd := "(Get-WinEvent -Logname System | Where-Object {$_.Message -Match '.*(" + service +
+			").*(stopped|terminated).*'} | Select-Object -Property TimeCreated -First 1 | foreach {$_.TimeCreated.ToString('R')} | Out-String).Trim()"
+		out, err := powershell(getTimeCreatedCmd)
+		if err != nil {
+			return time.Duration(0), err
+		}
+		if out == "" {
+			return time.Duration(0), fmt.Errorf("service time creation not found for %s", service)
+		}
+		out = strings.ReplaceAll(out, ",", "")
+		t, err := time.Parse(types.UptimeTimeLayout, out)
+		if err != nil {
+			return time.Duration(0), err
+		}
+		return time.Since(t), nil
+	}
+}
+
+// getRepairFunc returns the repair function based on the component.
+func getRepairFunc(hco *options.HealthCheckerOptions) func() {
+	// Restart-Service will stop and attempt to start the service
+	return func() {
+		powershell("Restart-Service", hco.Service)
+	}
+}
+
+// getHealthCheckFunc returns the health check function based on the component.
+func getHealthCheckFunc(hco *options.HealthCheckerOptions) func() (bool, error) {
+	switch hco.Component {
+	case types.KubeletComponent:
+		return func() (bool, error) {
+			httpClient := http.Client{Timeout: hco.HealthCheckTimeout}
+			response, err := httpClient.Get(types.KubeletHealthCheckEndpoint)
+			if err != nil || response.StatusCode != http.StatusOK {
+				return false, nil
+			}
+			return true, nil
+		}
+	case types.DockerComponent:
+		return func() (bool, error) {
+			if _, err := execCommand("docker.exe", "ps"); err != nil {
```

Comment:
Why do we use exec for docker, but powershell for crictl? Just for my education.

Reply:
I think I can change the crictl one to run directly instead of in a Powershell command. Originally, I had everything ran in a powershell, but per comments from Jeremy, I was told I can run docker.exe directly as opposed to using powershell. I overlooked that the crictl path is also an executable file. I'll test it and make updates once I test it works.

Reply:
updated